lib/std/reset_event.zig

```diff
@@ -0,0 +1,433 @@
+        assert(c.pthread_mutex_lock(&self.mutex) == 0);
+        defer assert(c.pthread_mutex_unlock(&self.mutex) == 0);
+
+        return self.state == IS_SET;
+    }
+
+    pub fn reset(self: *PosixEvent) bool {
+        assert(c.pthread_mutex_lock(&self.mutex) == 0);
+        defer assert(c.pthread_mutex_unlock(&self.mutex) == 0);
+
+        const was_set = self.state == IS_SET;
+        self.state = 0;
+        return was_set;
+    }
+
+    pub fn set(self: *PosixEvent, auto_reset: bool) bool {
+        assert(c.pthread_mutex_lock(&self.mutex) == 0);
+        defer assert(c.pthread_mutex_unlock(&self.mutex) == 0);
+
+        const had_waiter = self.state > IS_SET;
+        self.state = if (auto_reset) 0 else IS_SET;
+        if (had_waiter) {
+            assert(c.pthread_cond_signal(&self.cond) == 0);
+        }
+        return had_waiter;
+    }
+
+    pub fn wait(self: *PosixEvent, timeout: ?u64) ResetEvent.WaitError!bool {
+        assert(c.pthread_mutex_lock(&self.mutex) == 0);
+        defer assert(c.pthread_mutex_unlock(&self.mutex) == 0);
+
+        if (self.state == IS_SET)
+            return false;
+
+        var ts: os.timespec = undefined;
+        if (timeout) |timeout_ns| {
+            var timeout_abs = timeout_ns;
+            if (comptime std.Target.current.isDarwin()) {
+                var tv: os.darwin.timeval = undefined;
+                assert(os.darwin.gettimeofday(&tv, null) == 0);
+                timeout_abs += @intCast(u64, tv.tv_sec) * time.second;
+                timeout_abs += @intCast(u64, tv.tv_usec) * time.microsecond;
+            } else {
+                os.clock_gettime(os.CLOCK_REALTIME, &ts) catch unreachable;
+                timeout_abs += @intCast(u64, ts.tv_sec) * time.second;
+                timeout_abs += @intCast(u64, ts.tv_nsec);
+            }
+            ts.tv_sec = @intCast(@typeOf(ts.tv_sec), @divFloor(timeout_abs, time.second));
+            ts.tv_nsec = @intCast(@typeOf(ts.tv_nsec), @mod(timeout_abs, time.second));
+        }
+
+        var dummy_value: u32 = undefined;
+        var wait_token = @truncate(u32, @ptrToInt(&dummy_value));
+        self.state = wait_token;
+
+        while (self.state == wait_token) {
+            const rc = switch (timeout == null) {
+                true => c.pthread_cond_wait(&self.cond, &self.mutex),
+                else => c.pthread_cond_timedwait(&self.cond, &self.mutex, &ts),
+            };
+            // TODO: rc appears to be the positive error code making os.errno() always return 0 on linux
+            switch (std.math.max(@as(c_int, os.errno(rc)), rc)) {
+                0 => {},
+                os.ETIMEDOUT => return ResetEvent.WaitError.TimedOut,
+                os.EINVAL => unreachable,
+                os.EPERM => unreachable,
+                else => unreachable,
+            }
+        }
+        return true;
+    }
+};
+
+test "std.ResetEvent" {
+    // TODO
+    if (builtin.single_threaded)
+        return error.SkipZigTest;
```

Comment:
It looks like you did write code to handle this case, did you mean to enable this test case now?

Reply:
Oh, I see, the test spawns threads.